feat : Migrated Registration Fragment to MVVM

gururani-abhishek opened the pull request:
Fixes MOBILE-68

1. Created and added code to ViewModel and ViewModel factory for Registration Fragment MVVM Migration
2. Created a UserAuthRepository interface and it's implementation class to link ViewModel interactions with Business layer. 
3. Kept UI related code to View(RegistrationFragment) and shifted business logic from both View(RegistrationFragment) and Presenter(RegistrationPresenter) to ViewModel. 
4. Reformatted a few recurring instances of "binding" variable and fixed some errors which were a result of format issues with existing strings. 
5. Used DI to create newer modules.
6. Performed proper unit testing on viewModel and Repository using exiting framework -> jUnit and Mockito. 
7. Fixed build fail error due to unavailability of some libs. 


Please Add Screenshots If there are any UI changes.

https://github.com/openMF/mifos-mobile/assets/59947871/35dd1055-7fc8-4293-84b1-9fe68179350d


Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

gururani-abhishek committed:
feat : Migrated Registration Fragment to MVVM

gururani-abhishek commented:
Hey @jawidMuhammadi, please have a look at this. Please suggest that changes that you feel would make the code quality better.

gururani-abhishek commented:
> Hey @jawidMuhammadi, please have a look at this. Please also suggest the changes that you feel would make the code quality better.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/fragments/RegistrationFragment.kt:
```diff
@@ -41,195 +43,207 @@ class RegistrationFragment : BaseFragment(), RegistrationView {
-                    updatePasswordStrengthView(charSequence.toString())
+        viewModel = ViewModelProvider(this, viewModelFactory)[RegistrationViewModel::class.java]
+        with(binding) {
+            progressBar.visibility = View.GONE
+            passwordStrength.visibility = View.GONE
```
The initial value of View.GONE can be set in the XML file as well, no need to add here, remove it

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/fragments/RegistrationFragment.kt:
```diff
@@ -41,195 +43,207 @@ class RegistrationFragment : BaseFragment(), RegistrationView {
+            passwordStrength.setTextColor(str.color)
+            val mode = PorterDuff.Mode.SRC_IN
+            progressBar.progressDrawable?.setColorFilter(str.color, mode)
+            if (str.getText(context) == getString(R.string.password_strength_weak)) {
+                progressBar.progress = 25
+            } else if (str.getText(context) == getString(R.string.password_strength_medium)) {
+                progressBar.progress = 50
+            } else if (str.getText(context) == getString(R.string.password_strength_strong)) {
+                progressBar.progress = 75
+            } else {
+                progressBar.progress = 100
+            }
```
This section can be improved a lot but for now, at least can refactore this to use Kotlin's `when`

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/fragments/RegistrationFragment.kt:
```diff
@@ -41,195 +43,207 @@ class RegistrationFragment : BaseFragment(), RegistrationView {
+                val radioButton = rgVerificationMode.checkedRadioButtonId.let {
+                    root.findViewById<RadioButton>(it)
+                }
+                val payload = RegisterPayload().apply {
+                    accountNumber = etAccountNumber.text.toString()
+                    authenticationMode = radioButton?.text.toString()
+                    email = etEmail.text.toString()
+                    firstName = etFirstName.text.toString()
+                    lastName = etLastName.text.toString()
+                    mobileNumber =
+                        countryCodePicker.selectedCountryCode.toString() + etPhoneNumber.text.toString()
+                    if (etPassword.text.toString() != etConfirmPassword.text.toString()) {
+                        Toaster.show(root, getString(R.string.error_password_not_match))
+                        return
+                    } else {
+                        password = etPassword.text.toString()
+                    }
+                    password = etPassword.text.toString()
+                    username = etUsername.text.toString().replace(" ", "")
+                }
```
Creating a Payload object should be moved to data layer

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/ui/fragments/RegistrationFragment.kt:
```diff
@@ -41,195 +43,207 @@ class RegistrationFragment : BaseFragment(), RegistrationView {
+                Toaster.show(
+                    rootView,
+                    getString(R.string.error_validation_blank, getString(R.string.username)),
+                )
```
Could we please replace all this if-else madness with Kotlin's `when`?

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/viewModels/RegistrationViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+import org.mifos.mobile.models.register.RegisterPayload
+import javax.inject.Inject
+
+class RegistrationViewModel @Inject constructor(var dataManager : DataManager?) : ViewModel() {
```
Repository??? Please create a Repository interface and its implementation.

gururani-abhishek replied:
Jawid I created the RegistrationRepository Class without the RegistrationRepository Interface. Reason being we only have one source(i.e. remote source) for data fetching, so I found it redundant to create only one interface  file RemoteRegistrationRepositoryInterface and then implement it, as we don't have local or test source. If you still insist that we go with an interface, I'll do that. Thanks.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/viewModels/RegistrationViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+class RegistrationViewModel @Inject constructor(var dataManager : DataManager?) : ViewModel() {
+    private val compositeDisposables: CompositeDisposable = CompositeDisposable()
+
+    private var registrationResultSuccess = MutableLiveData<Boolean>()
+    val readRegistrationResultSuccess : LiveData<Boolean> get() = registrationResultSuccess
+    private var exceptionOnRegistration : Throwable? = null
+    val readExceptionOnRegistration : Throwable get() = exceptionOnRegistration!!
```
You could crate a UI state sealed class named RegistrationUiState

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/viewModels/RegistrationViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    fun isInputLengthInadequate(fieldText : String) : Boolean {
+        return fieldText.trim { it <= ' '}.length < 6
```
Is this in Kolinish way? :)

gururani-abhishek replied:
No it isn't, it was legacy code hence used it the same way. I'll replace _fieldText.trim { it <= ' ' }_ with **fieldText.trim()**

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/viewModels/RegistrationViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    fun registerUser(registerPayload: RegisterPayload?) {
+        dataManager?.registerUser(registerPayload)
```
Coroutines, please!

gururani-abhishek replied:
@jawidMuhammadi am I missing something, we were just performing MVVM migration right now?

jawidMuhammadi replied:
At some point, we'll need to migrate it. It was on our idea list and roadmap, and I hope you'll be able to implement it by the end of GSoC. I would have used the coroutine which is much easier for me. Howser, still it is all up to you guys on how to move forward and implement them. 

cc @PratyushSingh07

gururani-abhishek replied:
That's a great point Jawid. I'm up for it. I'll replace RxJava calls with coroutines. Great!!

gururani-abhishek replied:
Hey @jawidMuhammadi , I've made the changes except this one. Because I'm right now researching and learning how to use properly use coroutines in our project. Please have a look at this whenever you find time. Plus right now I didn't put created repository in a new Package because that would conflict with other contributor's PRs as well. For creating packages we can do a whole new PR.

gururani-abhishek committed:
feat : migration of Registration Fragment to MVVM.

gururani-abhishek committed:
feat : registration fragment migration to MVVM

1. added unit tests for registrationViewModel.

2. added unit tests for userAuthRepository

gururani-abhishek committed:
Merge branch 'mvvm_architecture' into registration-MVVM-migration

gururani-abhishek commented:
Hey @jawidMuhammadi, please have a look at this. Once this is done, I can work on other features that are in succession to it :>  Login, Update Passcode features. Thanks.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/viewModels/RegistrationViewModel.kt:
```diff
@@ -0,0 +1,66 @@
+import org.mifos.mobile.utils.RegistrationUiState
+import javax.inject.Inject
+
+class RegistrationViewModel @Inject constructor(private var userAuthRepositoryImp: UserAuthRepositoryImp?) : ViewModel() {
```
It should be UserAuthRepository, not the UserAuthRepositoryImp.

gururani-abhishek replied:
Done

jawidMuhammadi replied:
I was expecting something like this 🤦 :
```
class RegistrationViewModel @Inject constructor(
     val userAuthRepositoryImp: UserAuthRepository
) : ViewModel() {
}
```

gururani-abhishek replied:
my bad, fixed.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/repositories/UserAuthRepository.kt:
```diff
@@ -0,0 +1,11 @@
+
+interface UserAuthRepository {
+
+    fun registerUser(accountNumber: String?, authenticationMode: String?, email: String?,
```
Please reformat your code in this class an other classes you did changes.

gururani-abhishek replied:
done reformatting based on android studio guidelines.

gururani-abhishek committed:
fix : testing with build fail,

	added two libs to test around build fail :
	1. mockito-core
	2. mockito-android

gururani-abhishek committed:
feat : registration fragment MVVM migration

1. code reformat

2. injected userAuthRepository interface dependency instead of userAuthRepositoryImp.

3. making dataManager and userAuthRespository as non-nullable.
4. using val instead of var for dependencies.

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/repositories/UserAuthRepositoryImp.kt:
```diff
@@ -0,0 +1,34 @@
+import org.mifos.mobile.models.register.RegisterPayload
+import javax.inject.Inject
+
+class UserAuthRepositoryImp @Inject constructor(private var dataManager: DataManager?) :
```
Can we use `val` and make it not nullable? - remove the question mark, please

gururani-abhishek replied:
done, I have a question : 
Because we are making a dependency that's being injected by di framework as non-nullable are we sure that it'll always be non-nullabe? And could there ever be a possibility that it could be non-null?

jawidMuhammadi replied:
Yeah, it is guaranteed that it won't be null, if there is any issue with the dependency then a complete-time error will be thrown.

jawidMuhammadi replied:
`var` to `val`??? 😞

gururani-abhishek replied:
😅 done

jawidMuhammadi commented on app/src/main/java/org/mifos/mobile/viewModels/RegistrationViewModel.kt:
```diff
@@ -0,0 +1,80 @@
+import org.mifos.mobile.utils.RegistrationUiState
+import javax.inject.Inject
+
+class RegistrationViewModel @Inject constructor(private var userAuthRepositoryImp: UserAuthRepository?) :
```
Same here

gururani-abhishek replied:
done

jawidMuhammadi reviewed (changes requested):
Other parts of the code look good to me. Once the requested changes are done, I will merge the code.

gururani-abhishek committed:
Merge remote-tracking branch 'origin/mvvm_architecture' into registration-MVVM-migration

# Conflicts:
#	app/src/main/java/org/mifos/mobile/injection/component/ActivityComponent.kt
#	app/src/main/java/org/mifos/mobile/injection/component/ApplicationComponent.kt
#	app/src/main/java/org/mifos/mobile/ui/fragments/RegistrationFragment.kt

gururani-abhishek committed:
fix : Registration fragment MVVM migration

resolved conflicts

jawidMuhammadi reviewed (approved):
I think it looks good. Great job!
